onnxruntime/python/tools/quantization/calibrate.py

```diff
@@ -8,284 +8,224 @@
-                                           Calibraton will be done for them.
-        parameter black_nodes: nodes with these names will be force ignored by this
-                               calibration augmentation, no mather what's their op type.
-        parameter white_nodes: nodes with these names will be force to be calibration augmented.
-        return: augmented ONNX model
-    '''
-
-    added_nodes = []
-    added_outputs = []
-    for node in model.graph.node:
-        should_be_calibrate = ((node.op_type in quantization_candidates) and
-                               (node.name not in black_nodes)) or (node.name in white_nodes)
-        if should_be_calibrate:
-            input_name = node.output[0]
-            # Adding ReduceMin nodes
-            reduce_min_name = ''
-            if node.name != '':
-                reduce_min_name = node.name + '_ReduceMin'
-            reduce_min_node = onnx.helper.make_node('ReduceMin', [input_name], [input_name + '_ReduceMin'],
-                                                    reduce_min_name,
-                                                    keepdims=0)
-            added_nodes.append(reduce_min_node)
-            added_outputs.append(helper.make_tensor_value_info(reduce_min_node.output[0], TensorProto.FLOAT, ()))
-
-            # Adding ReduceMax nodes
-            reduce_max_name = ''
-            if node.name != '':
-                reduce_max_name = node.name + '_ReduceMax'
-            reduce_max_node = onnx.helper.make_node('ReduceMax', [input_name], [input_name + '_ReduceMax'],
-                                                    reduce_max_name,
-                                                    keepdims=0)
-            added_nodes.append(reduce_max_node)
-            added_outputs.append(helper.make_tensor_value_info(reduce_max_node.output[0], TensorProto.FLOAT, ()))
-    model.graph.node.extend(added_nodes)
-    model.graph.output.extend(added_outputs)
-    return model
-
-
-# Using augmented outputs to generate inputs to quantize.py
-
-
-def get_intermediate_outputs(model_path, session, inputs, calib_mode='naive'):
-    '''
-    Gather intermediate model outputs after running inference
-        parameter model_path: path to augmented FP32 ONNX model
-        parameter inputs: list of loaded test inputs (or image matrices)
-        parameter calib_mode: type 'naive' gives (ReduceMin, ReduceMax) pairs
+import abc
+
+class CalibrationDataReaderInterface(metaclass=abc.ABCMeta):
+    @classmethod
+    def __subclasshook__(cls,subclass):
+        return (hasattr(subclass,'get_next') and callable(subclass.get_next) or NotImplemented)
+
+    @abc.abstractmethod
+    def get_next(self) -> dict:
+        """generate the input data dict for ONNXinferenceSession run"""
+        raise NotImplementedError
+
+class ONNXCalibrater:
+    def __init__(self,
+                 model_path,
+                 data_reader,
```

Comment:
could you please to add a comment to indicate it should be the type of CalibrationDataReaderInterface

Reply:
actually, you can write like this data_reader:CalibrationDataReader

---
In reply to: [458339480](https://github.com/microsoft/onnxruntime/pull/4561#discussion_r458339480) [](ancestors = 458339480)